### ISSUE TITLE
unable to catch errors

### ISSUE THREAD

**Issue by chooin** [colinhacks/zod#3310]
I use promise, but I can't get the error in catch.

```ts
zod.number({
  required_error: 'title is required',
}).promise().parse(title).catch(() => {});
```

<img width="680" alt="image" src="https://github.com/colinhacks/zod/assets/10768858/777fb37d-086d-4f38-94db-130976a655c0">

**Comment by JacobWeisenburger** [comment]
Is this what you are looking for?

https://github.com/colinhacks/zod?tab=readme-ov-file#safeparse

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Comment by chooin** [comment]
@JacobWeisenburger Although this method can indeed solve the error reporting problem, I don't think it's the most elegant solution. Typically, users would retrieve error information in the catch block of the promise.

**Comment by shaharke** [comment]
@chooin Can you share the full code snippet including where you initialize `title`? My guess is that title is not a promise and the error you're getting is due to incorrect usage of the API.

**Comment by chooin** [comment]
@shaharke Yes, I just feel that this promise without a catch does not conform to user habits

**Comment by chooin** [comment]
If `ZodPromise` does not return a catch, is it more reasonable to use `PromiseLike`?

**Comment by shaharke** [comment]
@chooin I'm not sure I agree. the Purpose of the `.promise` function as far as I understand it is not to return a Promise or PromiseLike object when you call parse, but to validate that the input you're passing it is a Promise (see https://zod.dev/?id=promises). For the former you should call `.parseAsync()`  like so:

```
zod.number({
  required_error: 'title is required',
}).parseAsync(title).catch(() => {});
```

**Comment by colinhacks** [comment]
@chooin There's rarely a reason to use ZodPromise unless you're trying to properly type a `ZodFunction`. 

There are different *methods* for parsing async values, you don't need to define ZodPromise schemas to do what you're trying to do:

```ts
z.number({
  required_error: 'title is required',
}).parseAsync(new Promise(res => res('asdf'))).catch(() => {
  console.log('caught');
});
```

Let me know if I'm misunderstanding.

**Comment by chooin** [comment]
@shaharke @colinhacks Thank you for your answers, your design is pretty good, it's my misunderstanding.
